service/deployer/eventer/github/client.go

```diff
@@ -0,0 +1,163 @@
+package github
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	microerror "github.com/giantswarm/microkit/error"
+)
+
+const (
+	// deploymentUrlFormat is the string format for the GitHub
+	// API call to fetch Deployments.
+	deploymentUrlFormat = "https://api.github.com/repos/%s/%s/deployments"
+
+	// deploymentStatusUrlFormat is the string format for the
+	// GitHub API call to post Deployment Statuses.
+	deploymentStatusUrlFormat = "https://api.github.com/repos/%s/%s/deployments/%v/statuses"
+
+	// etagHeader is the header used for etag.
+	// See: https://en.wikipedia.org/wiki/HTTP_ETag.
+	etagHeader = "Etag"
+)
+
+// request makes a request, handling any metrics and logging.
+func (e *githubEventer) request(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", fmt.Sprintf("token %s", e.oauthToken))
+
+	resp, err := e.client.Do(req)
+	if err != nil {
+		return nil, microerror.MaskAny(err)
+	}
+
+	// Update rate limit metrics.
+	if err := updateRateLimitMetrics(resp); err != nil {
+		return nil, microerror.MaskAny(err)
+	}
+
+	return resp, err
+}
+
+// filterDeployments filters out any deployments we don't want,
+// such as deployments for other installations.
+func (e *githubEventer) filterDeployments(deployments []deployment) []deployment {
+	matches := []deployment{}
+
+	for _, deployment := range deployments {
+		if deployment.Environment == e.environment {
+			matches = append(matches, deployment)
+		}
+	}
+
+	return matches
+}
+
+// fetchNewDeploymentEvents fetches any new GitHub Deployment Events for the
+// given project.
+func (e *githubEventer) fetchNewDeploymentEvents(project string, etagMap map[string]string) ([]deployment, error) {
+	e.logger.Log("debug", "fetching deployments", "project", project)
+
+	url := fmt.Sprintf(
+		deploymentUrlFormat,
+		e.organisation,
+		project,
+	)
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, microerror.MaskAny(err)
+	}
+
+	// If we have an etag header for this project, then we have already
+	// requested deployment events for it.
+	// So, set the header so we only get notified of new events.
```

Comment:
How is the etag header correlated to the state of events? I understood the etag header was to disable the rate limiting. Is it disabling the rate limitting because we get the same response over and over again?